index.d.ts

```diff
@@ -1,12 +1,12 @@
-import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
+import { IconDefinition, IconName, IconPrefix } from '@fortawesome/fontawesome-svg-core';
 import { DefineComponent } from 'vue'
 
 interface FontAwesomeIconProps {
   border?: boolean
   fixedWidth?: boolean
   flip?: 'horizontal' | 'vertical' | 'both'
-  icon: object | Array<string> | string | IconDefinition
-  mask?: object | Array<string> | string
+  icon: [IconPrefix, IconName] | IconName | IconDefinition
+  mask?: [IconPrefix, IconName] | IconName
```

Comment:
@Mensix --- Would/could this break backwards compatibility ?

Reply:
Oh, good question! I tested the change with the Nuxt framework - obviously VS Code gave autocompletion properly. Once given wrong icon name, an TS related error TypeScript appears, however... it means nothing at all, as typing errors are ignored both in the dev and build mode. Seems like the case also for Vite and vue-cli according to their docs, unless using vue-tsc utility. 

So 🤔, I will have a further look on it tomorrow.

Reply:
I just noticed that there is **a breaking change** -- string syntax wouldn't be further supported. If we try to extend types by it:

```typescript
icon: [IconPrefix, IconName] | IconName | IconDefinition | [CssStyleClass, `fa-${IconName}`]
mask?: [IconPrefix, IconName] | IconName | [CssStyleClass, `fa-${IconName}`]
```

then (at least VSC) gives lots of pointless options, like ```fa-duotone fa-facebook```. But!

We could also totally make the string option not supported by autocompletion, yet the rest would still work, by using, well, weird TypeScript hacks: https://github.com/microsoft/TypeScript/issues/29729